### PR TITLE
[REBASE & FF] Add GetVariable Interface for AdvLogger and Support AdvLogger in PEI

### PR DESCRIPTION
**Pull request by os-d**
## Description

Add the GetVariable hook used by AdvancedLogger and the PlatformBlob used in PEI. Cherry-picked from 26e6699b15
c18444f829 in release/202311.

GetVariableHook
--
One of the ways to fetch the AdvLogger log from an external source (general the host OS) is through a GetVariable call. These are not real variables and instead there is a hook that sees the AdvLogger variable names and gets the relevant log data from the in-memory log. This requires a change in basecore to add the hook to the GetVariable call.

PlatformBlob
--
The other change required in basecore for AdvLogger is adding the pointer to the log to the PEI Core instance structure so that in a pre-memory environment, the AdvLogger info structure (and log) can be used. This requires adding a new field to the structure defined in basecore.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

From 2311

## Integration Instructions

N/A.

### CODE REVIEW

**Review by kenlautner — APPROVED** [review]

File: MdeModulePkg/Core/Pei/PeiMain.h
Comment (makubacki) [inline MdeModulePkg/Core/Pei/PeiMain.h:1]: Doesn't most of this center around the GetVariable() path for advanced logger? If so, maybe make that more clear in the PR title and description for the release notes.

Comment (os-d) [reply MdeModulePkg/Core/Pei/PeiMain.h:1]: Yes, will do. I was lazy on my initial PR descriptions, cleaning them up now.

File: MdeModulePkg/Include/Library/AdvLoggerAccessLib.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/** @file
+  SMM GetVariable Implementation to retrieve the AdvLogger memory.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.<BR>
```
Comment (makubacki) [inline MdeModulePkg/Include/Library/AdvLoggerAccessLib.h]:
```suggestion
  Copyright (c) Microsoft Corporation.

```

File: MdeModulePkg/Universal/Variable/RuntimeDxe/VariableSmm.c [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 Copyright (c) 2010 - 2019, Intel Corporation. All rights reserved.<BR>
 Copyright (c) 2018, Linaro, Ltd. All rights reserved.<BR>
+Copyright (c) Microsoft Corporation.  All rights reserved.<BR>         // MU_CHANGE
```
Comment (makubacki) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/VariableSmm.c]:
```suggestion
Copyright (c) Microsoft Corporation.        // MU_CHANGE

```

File: MdeModulePkg/Library/AdvLoggerAccessLibNull/AdvLoggerAccessLib.inf [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+[Packages]
+  MdePkg/MdePkg.dec
+  MdeModulePkg/MdeModulePkg.dec
+
+[LibraryClasses]
+
+[Guids]
+  gAdvLoggerAccessGuid
+
+[Protocols]
```
Comment (makubacki) [inline MdeModulePkg/Library/AdvLoggerAccessLibNull/AdvLoggerAccessLib.inf]: If you're going to do another push, can you please drop the empty sections from this file?

File: MdeModulePkg/Library/AdvLoggerAccessLibNull/AdvLoggerAccessLib.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/** @file
+  SMM GetVariable NULL Implementation to retrieve the AdvLogger memory using
+  GetVariable.
+
+
+  Copyright (c) Microsoft Corporation. All rights reserved.<BR>
```
Comment (makubacki) [inline MdeModulePkg/Library/AdvLoggerAccessLibNull/AdvLoggerAccessLib.c]:
```suggestion
  Copyright (c) Microsoft Corporation.

```

File: MdeModulePkg/Include/Library/AdvLoggerAccessLib.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/** @file
+  SMM GetVariable Implementation to retrieve the AdvLogger memory.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __ADV_LOGGER_ACCESS_LIB_H__
+#define __ADV_LOGGER_ACCESS_LIB_H__
```
Comment (makubacki) [inline MdeModulePkg/Include/Library/AdvLoggerAccessLib.h]:
```suggestion
#ifndef ADV_LOGGER_ACCESS_LIB_H_
#define ADV_LOGGER_ACCESS_LIB_H_

```

File: MdeModulePkg/Include/Library/AdvLoggerAccessLib.h (hunk 1 above)
Comment (os-d) [reply MdeModulePkg/Include/Library/AdvLoggerAccessLib.h]: Will fix all the licenses.

**Review by makubacki — APPROVED** [review]

**Pull request merged**